### DISCUSSION TITLE
Swap channels of input

### DISCUSSION THREAD

**Discussion by jakub-zawislak** [ExistentialAudio/BlackHole#13] [Feature Requests]
I didn't found such functionality in readme, but maybe I just don't know how to use this program. 

My use case:
I have a Scarlett Solo with instrument input at channel 2. I want to use it in the Rocksmith 2014, which listens only to channel 1. So I need to swap channels of input.

I can do it in the Loopback, but for the price of license, I could buy a more expensive Scarlett 2i2 which also solves this problem.

**Comment by devinroth** [comment]
It’s definitely possible but not implemented. I can add it to the feature requests. If you want to dig into the code you could modify it to fit your needs. 

Devin

**Reply by revilo464** [reply]
I would like to try to poke in the source code. :) Any tips?

**Reply by devinroth** [reply]
Look at BlackHole_DoIOOperation.

**Reply by revilo464** [reply]
So is creating an effective UI the biggest issue standing in the way of re-directing channels?

**Reply by devinroth** [reply]
I don't think it makes any sense doing it in the driver. Possible though.

**Comment by jeffp123** [comment]
I have a similar use-case.

I have a Steinberg UR22 and I want to use it with a direct-in bass in the right channel and combine it with some music that I want to play along with (from Spotify or a web browser, etc.) and to record that with QuickTime Player.

I can do that currently with SoundFlower (and presumably Blackhole too), but the bass is only in the right channel.

With Loopback, I am able to send the right channel to the left/right output channels. Looks like this:
<img width="610" alt="Screen Shot 2019-11-13 at 1 14 44 AM" src="https://user-images.githubusercontent.com/4606444/68737979-09cdf980-05b3-11ea-83d0-4a9f23d115aa.png">

But, I concur with with @jakub-zawislak that it is not worth the $99 to only use that one feature.

I could conceivably do this in a DAW, but I am trying to rely entirely on the system audio options and record direct to QuickTime, without having to mix things after the fact.

I think that the ability to pipe in audio from any channel of any device would be a great feature.

**Reply by fuguesoft** [reply]
Yes yes yes!
This is something that would be huge

**Comment by PatTheMav** [comment]
A +1 from me for what @chaimpeck wrote - I currently need to route two channels of an external audio device to two seperate loopback devices as the target software doesn't allow channel selection (and downmixes everything to stereo per device).

I could compile blackhole twice with different UUIDs to get those two loopbacks (something that _might_ be a great built-in feature: blackhole "clones"), but I'd still need some way to route a single channel from device A to a seperate blackhole device.

**Comment by thomasberglund** [comment]
@devinroth Awesome work with BlackHole, and thanks for sharing it in the open with the rest of the world. Amazing.

Skype has the same issue as other apps mentioned here. It only listens to channel 1/2, with multi-channel audio interfaces – hence no audio from other inputs via BlackHole in Skype.

**Comment by krono-i2** [comment]
+1 I need this functionality for Serato DJ Pro.

**Comment by williammuff** [comment]
I second the Serato DJ comment!

**Comment by Teufeuleu** [comment]
+1 !

**Comment by reneotten** [comment]
+1

**Comment by jts3k** [comment]
+1

**Comment by rmonge** [comment]
+1

**Comment by nathanpthousand** [comment]
+1 the Zoom R24 has no phantom power to channels 1 and 2 and is useless as an interface for almost all streaming applications.

**Comment by dexterslabor** [comment]
+1 Traktor DJ

**Comment by andregriffin** [comment]
Channel matrix routing like this would be amazing. And I'd think the latency here would be less than Loopback doing the same thing.

For example, I'm using an RME UFX with 30in/30out. The UFX has it's own set of routing and loopback features, however I need to pick up inputs 17/18 from the UFX and route them to Blackhole 1/2, for use in applications that don't support multichannel audio cards. 

There's seemingly no reliable way to do this without using a $99 app like Loopback or something like LadioCast (which is unreliable for some reason).

This would be a huge feature, especially for those needing to submix audio for streaming to OBS with the least amount of latency.

**Comment by julientaq** [comment]
Hi!

Got a zoom r16 with 8 channel, and the 48v for my mic is on channel 5/6. I would love to be able to send the audio through the web browser for Big blue button / jisti / ~~zoom~~ / etc. sessions.

Thanks for the great work!

Ju

**Comment by zfogg** [comment]
+1

**Comment by EduardsE** [comment]
+1

**Comment by revilo464** [comment]
Hey @devinroth, I figured out a way to re-route audio channels using your driver but using Reaper. I did dive into the driver source code and tried to determine a way to accomplish it at the driver-level but it would require a lot more in-depth reading and understanding of the AudioServerPlugInDriver interface. I would somehow have to request the output of another device in the driver.

I am one of those who was looking to find a way to play Rocksmith on Mac (which only listens on channel 1 from a device that has the name "Rocksmith USB Guitar Adapter"). After playing with the driver a bit, I instead just renamed the device to the required name in the BlackHole.h file and then built the driver without signing. I then used Reaper to route audio between channels in an aggregate device containing my audio interface and the re-named BlackHole, sending line in from my audio interface's channel 2 to channel 1 on the BlackHole device.

This works successfully and the game now accepts input, however, I would like to share my findings with the community over at the /r/Rocksmith subreddit and all may not have Xcode installed. Would I be allowed to share my renamed and un-signed BlackHole driver with them? If that's not acceptable, I can instead make a guide explaining how to build and rename the driver using Xcode.

EDIT: I should note that creating an aggregate device with the name "Rocksmith USB Guitar Adapter" used to work, but now it seems like it needs a "full device" with the name which requires building the driver with the required name.

**Reply by yozw** [reply]
Can you use the setting "MonoToStereoChannel=1" to the [Audio] section of your Rocksmith.ini file to switch to channel 2? https://www.reddit.com/r/rocksmith/comments/sihbcd/rocksmith_asio_on_mac/

**Comment by devinroth** [comment]
Yes you can share it. But also share the instructions. That will ensure that everything remains open source.

**Comment by gdhgdhgdh** [comment]
Another common use case for DJs - I have a Pioneer DJM-S9 which sends 8 channels of audio into the Mac. I'm only interested in channels 5 and 6, and would love to be able to reroute these to stereo left+right on the Blackhole 2ch device so that OBS can pick it up.

**Comment by matthew-manuel** [comment]
+1  I am having a similar issue as the other DJ users: trying to route the Master Output channels (5 & 6) from my Numark Scratch Mixer, but Twitch Studio assumes that the 6 channels from my mixer are L, R, C, Sub, Back L & R (i.e. a 5.1 surround layout when those respective channels are actually Deck 1 L & R, Deck 2 L & R, and Master L & R.  

I also got it working with Loopback but also don't feel this warrants the $100 USD price (as much as I love the UI.

**Reply by zfogg** [reply]
we should really make this nice and usable for DJs tbh.. if you're reading this you probably both dj and code haha..

**Comment by huseyz** [comment]
In case anyone is looking for an alternative; free solution for Rocksmith here is my setup with EVO4 which has the same issue.
1. Compile Blackhole with device name as "Rocksmith USB Guitar Adapter"
2. Use [LadioCast](https://apps.apple.com/us/app/ladiocast/id411213048?mt=12) to route audio

**Reply by yozw** [reply]
In case it's helpful (I've been Googling for two hours), I just found this: https://www.reddit.com/r/rocksmith/comments/sihbcd/rocksmith_asio_on_mac/. You can use the setting "MonoToStereoChannel=1" in the [Audio] section of your Rocksmith.ini file to switch to channel 2.

**Comment by fbodlee** [comment]
Another Use Case I found for this feature: Tascam's Portacapture X8 by default sends a stereo mix of its 6 inputs to its outputs 1 and 2 when in audio interface mode. The 2 USB output channels have to be mapped to two of hose 6 inputs to be audible on the headphones connected to the X8. Now the problem is that every video conferencing software I know of uses inputs 1 and 2 as the microphone when selecting a multi-channel audio interface (e.g. Google Meet). Since the computer audio gets mixed into inputs 1 and 2 of the Tascam, the computer audio always gets routed into the video conference. I wish Tascam would let us remap the outputs but until then, remapping them as described in this feature request seems to be the only possible solution for that issue.
Thanks for your contributions! 🥳

**Comment by TimSusa** [comment]
Hey Guys, Why doing the same work twice? I was searching a while and like to share with you. At the moment I built the app and it is running now. Looks very promising to me as I already have it running with BlackHole at the input and aggregated Device at the out put. You have a Graph Editor in so you can patch kind of the craziest stuff arround. It is a JUCE Project, so it is really like built on top of giants. Check that out. https://github.com/kushview/Element/blob/master/docs/building.md

**Reply by sivert-io** [reply]
# This!! :fire::fire:

I really didn't want to bother with a huge application, so Element is perfect.

### _A caveat_
- _You need to have Element running for the routing to work. You could simply minimize it and forget about it, but keep this in mind._

### _My use case_
_I use the **Roland VT-4** (6 channels in, 6 channels out) as my daily-driver Audio Interface, and the problem on MacOS is that the channels are all combined. This means the output audio sent to my interface is routed back into the microphone input. If I'm in a Teams meeting it means the participants will hear themselves. **Element + blackhole fixes it!**_

## Here's how I did it 🎉 
1. Install Element (https://github.com/kushview/element/releases download latest version. For mac, download the matching name: element-osx-x.xx.x.dmg). **You do not need to install Plugins**.
2. Install Blackhole. Channels don't really matter, 2ch should be fine but 16ch is also okay. It depends on how many channels you wish to route.
3. Open Element and go to Preferences (Cmd + ,), click the **Audio** tab and set _Output_ as **BlackHole 2ch** and _Input_ as your preferred input-device. This could be e.g. your microphone, or you can create an [aggregated device](https://support.apple.com/en-us/102171) and use it as input.

![image](https://github.com/ExistentialAudio/BlackHole/assets/6785315/1eb3cf70-466b-43d4-b422-d04d7300c583)


5. Close the preferences-window and change the channels to your preferences. Here's what my setup looks like:

![CleanShot 2024-06-11 at 14 01 45](https://github.com/ExistentialAudio/BlackHole/assets/6785315/88651220-b7b3-4f76-b8ab-11fbdc0a2317)

6. Set the default input-device as **BlackHole 2ch** in your audio settings.
## 6. Profit 🎉

**Comment by gunku** [comment]
+1 from me
I am using Audient Evo card where 1st channel is shared betweem mic and guitar input (Ig guitar is plugged in it overtakes Mic). I'd like to have Mic in Channel 2 so that I could have both of them in pro apps. However most non-pro conference call apps listen to channel one only, so option to swap them in BlackHole would be great!

**Comment by mat-kalinowski** [comment]
+1

**Comment by LaKing** [comment]
+1

**Comment by map-lo** [comment]
+1

**Comment by nilmic** [comment]
+1
I am using Presonus Quantum 2626 and currently I need GigPerformer running to route CH 8 input to BlackHole Input 1 & 2 for Teams meetings.
A simple routing configuration file in BlackHole is all I need.

**Comment by dpcavalcanti** [comment]
+1
